Attach dashcard's viz settings to the card's :data :viz-settings key

adam-james-v opened the pull request:
Closes: https://github.com/metabase/metabase/issues/35437
partially fixes: https://github.com/metabase/metabase/issues/33727 (solves column order/hiding but not other formatting issues yet)

The linked issues had similarities: csv/xlsx exports did not respect column settings (like hidden columns). 

In the case of dashboard subscription attachments, there were 2 issues:
 - the dashcard viz settings were not being incorporated into the call to `metabase.query-processor.card/run-query-for-card-async`, which is ultimately run for each dashcard (excluding virtual cards)
 - even when they were merged into the card viz-settings, they weren't normalized and so downstream fns didn't incorporate the settings, because the keys weren't actually in the viz-settings (this is an opportunity for schemas + cleaning up the viz-settings throughout the codebase in future PRs)

Why merge the dashcard viz settings and card viz settings? When the query processor runs the `process-viz-settings` middleware, it is unaware of any dashcards related to the cards, and no dashcard-ids or data get passed through the qp, so the settings are unavailable unless they're merged. In this PR they get merged onto the `:viz-settings` key of the query map that gets passed through the query processor. (Another thing worth cleaning up is sensible and consistent naming (viz-settings vs. visualization_settings for example)

In the case of the embedding, there were 2 issues:
 - the middleware map was not properly being passed through the embed API to the `run-query-for-card-async` function
 - similarly to above, the viz-settings were not normalized, so column order/hidden settings were not respected

## After this PR
<img width="1319" alt="image" src="https://github.com/metabase/metabase/assets/21064735/9aa0c9dd-eafe-4af5-9e5f-61392a957a9d">

The exports/downloaded csv (and xlsx) looked like the table on the left before. After, they look like on the right. (Hidden columns are not visible and the column order set in the dashboard is also applied properly).

There are now 2 additional tests:
 - the dashboard subscription test namespace now sends an email with .csv attachment and looks at the contents of the csv to assert that not all columns are shown
 - the api/embed test namespace adds a function that similarly asserts that column order / hidden state is respected in the csv exports.

adam-james-v committed:
Attach dashcard's viz settings to the card's :data :viz-settings key

This flows through to the export format fn `metabase.email.messages/stream-api-results-to-export-format`
Where it is then properly used.

adam-james-v committed:
Added test that gets the contents of .csv attachments

Not sure yet why it's not working. Locally, with a real db id it works as expected, so it's on the right track.

adam-james-v committed:
Create a working test.

adam-james-v committed:
Card qp merges dashcard viz when it exists. Viz middleware normalizes card/dashcard viz-settings

adam-james-v commented on src/metabase/api/embed.clj:
```diff
@@ -283,7 +283,8 @@
      :context       :embedded-dashboard
-     :constraints   constraints)))
+     :constraints   constraints
+     :middleware    middleware)))
```
Actually pass the middleware map to the query processor.

adam-james-v commented on src/metabase/query_processor/card.clj:
```diff
@@ -191,33 +191,38 @@
   options."
   [card-id export-format
-   & {:keys [parameters constraints context dashboard-id middleware qp-runner run ignore_cache]
+   & {:keys [parameters constraints context dashboard-id dashcard-id middleware qp-runner run ignore_cache]
```
pass in a dashcard-id here. May be nil. It'll be used to get and merge viz-settings from the dashcard (if there are any).

adam-james-v commented on src/metabase/query_processor/card.clj:
```diff
@@ -191,33 +191,38 @@
+                      (^:once fn* [query info]
+                       (qp.streaming/streaming-response [{:keys [rff context]} export-format (u/slugify (:card-name info))]
+                         (qp-runner query info rff context))))
+        dash-viz  (when (not= context :question)
```
get the dashcard's viz-settings. Might be an empty map.

adam-james-v commented on src/metabase/query_processor/card.clj:
```diff
@@ -191,33 +191,38 @@
+                                                  :dataset :result_metadata :visualization_settings]
+                                                 :id card-id))
+        query     (-> (query-for-card card parameters constraints middleware {:dashboard-id dashboard-id})
+                      (update :viz-settings (fn [viz] (when (or (seq viz) (seq dash-viz))
```
The `query` that will be run by the query processor. Has a :viz-settings key which may be nil or may have the card's viz settings. Only update the key if either card or dashcard viz settings exist and aren't empty. This is because the private fn in this namespace `viz-settings` will first look for `:viz-settings` on the query. Only if that's nil will it instead look up the card's viz settings from the qp store. So, an empty map short circuits this logic and causes other tests to fail.

markbastian replied:
Do you need the or logic in the update clause? I think you can safely update with merge as in the following examples:
```clojure
(update {} :viz-settings merge {:abc 123})
=> {:viz-settings {:abc 123}}
(update nil :viz-settings merge {:abc 123})
=> {:viz-settings {:abc 123}}
(update {:viz-settings {:abc "FOO"}} :viz-settings merge {:abc 123})
=> {:viz-settings {:abc 123}}
(update {} :viz-settings merge nil)
=> {:viz-settings nil}
(update {:viz-settings nil} :viz-settings merge nil)
=> {:viz-settings nil}
(update {:viz-settings {}} :viz-settings merge nil)
=> {:viz-settings {}}
```

adam-james-v replied:
The reason I did this  was for the case where the card's viz settings were not nil but the dashcard viz settings were empty {} The logic in the middleware was such that if you pass an empty map in, it won't bother to look up card viz settings from the qp store. So, tests started failing. 

I've changed the logic in the middleware and removed the (or ...) here.

markbastian commented on test/metabase/dashboard_subscription_test.clj:
```diff
@@ -881,7 +885,55 @@
+    (mt/with-fake-inbox
+      (mt/with-temp [Card {card-id :id :as c} (pulse.test-util/checkins-query-card {:breakout [!day.date]})
+                     Dashboard     {dash-id :id} {:name "just dash"}]
+        (let [viz {:table.columns (mapv metadata->field-ref (:result_metadata c) [true false])}]
```
Might be worth a comment as to what these viz settings do so the reader knows what the expected applied changes are.

adam-james-v replied:
Good call, added.

adam-james-v committed:
Addressed feedback.

 - if viz settings map is empty in process viz settings middleware, use qp store.
 - now we don't have to do the (or ...) logic anywhere else
 - added comment in the test to try clarify what the 'viz' is setting up

markbastian reviewed (approved):
Looks good. 🚢 🇮🇹
